atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -231,40 +243,44 @@ def get_cluster_info(self, cluster, platform):
         osbs = OSBS(conf, conf)
         try:
             current_builds = self.get_current_builds(osbs)
-        except OsbsException as e:
+        except OsbsException:
             # If the build is canceled reraise the error
-            if isinstance(e.cause, BuildCanceledException):
-                raise e
-
-            self.log.exception("Error occurred while listing builds on %s",
-                               cluster.name)
-            return ClusterInfo(cluster, platform, osbs, self.UNREACHABLE_CLUSTER_LOAD)
+            raise
 
         load = current_builds / cluster.max_concurrent_builds
         self.log.debug('enabled cluster %s for platform %s has load %s and active builds %s/%s',
                        cluster.name, platform, load, current_builds, cluster.max_concurrent_builds)
         return ClusterInfo(cluster, platform, osbs, load)
 
-    def choose_cluster(self, platform):
-        config = get_config(self.workflow)
-        clusters = [self.get_cluster_info(cluster, platform) for cluster in
-                    config.get_enabled_clusters_for_platform(platform)]
-
-        if not clusters:
-            raise RuntimeError('No clusters found for platform {}!'
-                               .format(platform))
+    def get_clusters(self, all_clusters, cluster_fails, platform, retry_at):
+        ''' return clusters sorted by load '''
+
+        clusters = []
+        while all_clusters and not clusters:
+            for cluster in all_clusters:
+                if cluster_fails[cluster.name] >= self.max_cluster_fails:
+                    all_clusters.remove(cluster)
```

Comment:
This logic, of removing clusters once they have failed the maximum number of times, seems to be both here and in select_and_start_cluster. Is it needed in both places?

Reply:
The only way I can think of is to pull the do_worker_build logic into select clusters or pull select clusters into the main loop. Both would make a rather large and hard to understand loop.

Reply:
What about making a class to handle the retry context? e.g.
```
class ClusterRetryContexts(object):
    def __init__(self, clusters, max_cluster_fails):
        self.clusters = list(clusters)
        self.retry_at = {}
        self.failed = {}
        self.fails = Counter()
        self.max_cluster_fails = max_cluster_fails

    def in_retry_wait(self, cluster):
        if cluster.name in self.retry_at:
            return datetime.datetime.now() < self.retry_at[cluster.name]

        return False

    def try_again_later(self, cluster, seconds):
        name = cluster.name
        self.fails[name] += 1
        if self.fails[name] >= self.max_cluster_fails:
            self.failed[name] = True
        else:
            self.retry_at[name] = (datetime.datetime.now() +
                                   timedelta(seconds=seconds))

        return not self.failed[name]

    def wait_for_any(self):
        time_until_next = (min(retry_at.values()) -
                           datetime.datetime.now().seconds)
        time.sleep(max(timedelta(seconds=0), time_until_next))
```

Then you'd be able to do this sort of thing:
```
clusters = []
while all_clusters and not clusters:
    if retry_info.in_retry_wait(cluster):
        continue
    try:
        ...
    except OsbsException:
        if not retry_info.try_again_later(cluster, self.find_cluster_retry_delay):
            all_clusters.remove(cluster)
    if not clusters:
        retry_info.wait_for_any()
```

and similarly in select_and_start_cluster().